src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java

```diff
@@ -152,7 +165,8 @@ protected TransportRequestOptions transportOptions() {
     }
 
     protected boolean retryPrimaryException(Throwable e) {
-        return TransportActions.isShardNotAvailableException(e);
+        return e.getClass() == RetryOnPrimaryException.class
```

Comment:
should we subclass `IllegalIndexShardStateException`  instead of `IndexShardException` then we don't need to change `retryPrimaryException`